strawberryfields/backends/bosonicbackend/bosoniccircuit.py

```diff
@@ -0,0 +1,833 @@
+        """
+        if modes is None:
+            modes = self.get_modes()
+        alpha = np.zeros(len(modes))
+        fidelity = self.fidelity_coherent(alpha, modes=modes)
+        return fidelity
+
+    def parity_val(self, modes=None):
+        r"""Returns the expectation value of the parity operator, which is the
+        value of the Wigner function at the origin.
+
+        Args:
+            modes (list or None): modes to use for parity calculation
+        """
+        if modes is None:
+            modes = list(range(self.nlen))
+        # Sort by (q1,p1,q2,p2,...)
+        mode_ind = np.sort(np.append(2 * np.array(modes), 2 * np.array(modes) + 1))
+        exp_arg = np.einsum(
+            "...j,...jk,...k",
+            self.means[:, mode_ind],
+            np.linalg.inv(self.covs[:, mode_ind, :][:, :, mode_ind]),
+            self.means[:, mode_ind],
+        )
+        weighted_exp = (
+            np.array(self.weights)
+            * np.exp(-0.5 * exp_arg)
+            / np.sqrt(np.linalg.det(self.covs[:, mode_ind, :][:, :, mode_ind]))
+        )
+        parity = np.sum(weighted_exp) * (self.hbar / 2) ** len(modes)
+        return parity
+
+    def loss(self, T, k):
+        r"""Implements a loss channel in mode k.
+
+        Args:
+            T (float between 0 and 1): loss amount is \sqrt{T}
+            k (int): mode that loses energy
+        """
+
+        if self.active[k] is None:
+            raise ValueError("Cannot apply loss channel, mode does not exist")
+
+        X = np.sqrt(T) * np.identity(2)
+        Y = self.hbar * (1 - T) * np.identity(2) / 2
+        X2, Y2 = self.expandXY([k], X, Y)
+        self.apply_channel(X2, Y2)
+
+    def thermal_loss(self, T, nbar, k):
+        r"""Implements the thermal loss channel in mode k.
+
+        Args:
+            T (float between 0 and 1): loss amount is \sqrt{T}
+            nbar (float): mean photon number of the thermal bath
+            k (int): mode that undegoes thermal loss
+        """
+        if self.active[k] is None:
+            raise ValueError("Cannot apply loss channel, mode does not exist")
+        X = np.sqrt(T) * np.identity(2)
+        Y = self.hbar * (1 - T) * (2 * nbar + 1) * np.identity(2) / 2
+        X2, Y2 = self.expandXY([k], X, Y)
+        self.apply_channel(X2, Y2)
+
+    def init_thermal(self, nbar, mode):
+        r"""Initializes a state of mode in a thermal state with the given population.
+
+        Args:
+            nbar (float): mean photon number of the thermal state
+            mode (int): mode that get initialized
+        """
+        self.thermal_loss(0.0, nbar, mode)
+
+    def is_vacuum(self, tol=0.0):
+        r"""Checks if the state is vacuum by calculating its fidelity with vacuum.
+
+        Args:
+            tol (float): the closeness tolerance to fidelity of 1
+        """
+        fid = self.fidelity_vacuum()
+        return np.abs(fid - 1) <= tol
+
+    def measure_dyne(self, covmat, indices, shots=1):
```

Comment:
(Though saw that such a renaming would break the convention with the Gaussian `Circuit`, maybe worth considering)
```suggestion
    def measure_dyne(self, covmat, modes, shots=1):
```

Reply:
Done.